Bot.Core/TownMaintenance.cs

```diff
@@ -0,0 +1,102 @@
+
+        private readonly ICallbackScheduler<Queue<TownKey>> m_callbackScheduler;
+        private readonly ITownDatabase m_townDatabase;
+        private readonly IBotClient m_botClient;
+        private readonly IDateTime m_dateTime;
+        private readonly IShutdownPreventionService m_shutdownPrevention;
+
+        private readonly TaskCompletionSource m_shutdownTcs = new();
+        private bool m_shutdownRequested = false;
+        private bool m_queueProcessing = false;
+
+        List<Func<TownKey, Task>> m_startupTasks = new();
+
+        public TownMaintenance(IServiceProvider sp)
+        {
+            sp.Inject(out m_dateTime);
+            sp.Inject(out m_townDatabase);
+            sp.Inject(out m_botClient);
+            sp.Inject(out m_shutdownPrevention);
+
+            m_shutdownPrevention.ShutdownRequested += (s, e) =>
+            {
+                m_shutdownRequested = true;
+                if (!m_queueProcessing)
+                {
+                    m_shutdownTcs.TrySetResult();
+                }
+            };
+            m_shutdownPrevention.RegisterShutdownPreventer(m_shutdownTcs.Task);
+
+            var callbackFactory = sp.GetService<ICallbackSchedulerFactory>();
+            m_callbackScheduler = callbackFactory.CreateScheduler<Queue<TownKey>>(ProcessQueue, TimeSpan.FromMinutes(1));
+
+            m_botClient.Connected += async (o, args) => await RunMaintenance();
+        }
+
+        private async Task ProcessQueue(Queue<TownKey> towns)
+        {
+            m_queueProcessing = true;
+            Serilog.Log.Information("TownMaintenance: {townCount} towns remain...", towns.Count());
+
+            int numSent = 0;
+            while (numSent < NUM_TOWNS_PER_CALLBACK && towns.Count > 0)
+            {
+                var townKey = towns.Dequeue();
+
+                Serilog.Log.Information("TownMaintenance: Processing town {townKey}", townKey);
+
+                foreach (var task in m_startupTasks)
+                {
+                    await task(townKey);
+                }
+                numSent++;
+            }
+
+            if (towns.Count > 0)
+            {
+                if (m_shutdownRequested)
+                    m_shutdownTcs.TrySetResult();
+                else
+                {
+                    Serilog.Log.Information("TownMaintenance: {townCount} towns remain, scheduling callback...", towns.Count);
+                    DateTime nextTime = m_dateTime.Now + TimeSpan.FromMinutes(MINUTES_PER_CALLBACK);
+                    m_callbackScheduler.ScheduleCallback(towns, nextTime);
+                }
+            }
+            else
+            {
+                Serilog.Log.Information("TownMaintenance: maintenance complete!");
+            }
+
+            m_queueProcessing = false;
+        }
+
+        public void AddMaintenanceTask(Func<TownKey, Task> startupTask)
+        {
+            m_startupTasks.Add(startupTask);
+        }
+
+        public async Task RunMaintenance()
```

Comment:
Is this actually used for anything? I think it's probably just unnecessary.

There's also a major problem with it, which is that it will result in 2 competing scheduled callbacks, both processing independent queues, but both manipulating member vars like `m_queueProcessing` and `m_shutdownTcs`.

The reason it will result in 2 competing queues is because the CallbackScheduler key is `Queue<TownKey>` which will be unique for:
* the constructor-spawned callback
* the process call here
* the callback spawned to continue work from `ProcessQueue`

Reply:
We should discuss this in person because I don't follow what you're saying here at all... the only place I schedule a callback is in `ProcessQueue` if items are remaining... each time that queue will be unique, yes.

Oh - are you saying we don't want some external party to call `RunMaintenance` until the existing run is done? I could guard that ... and yes, at the moment nobody uses this public `RunMaintenance` call at all, so we could just remove it

Reply:
Yes the confusion on my end is that it's public. If something else calls `RunMaintenance` (now `RunMaintenanceAsync` I think) there will be problems.